### PR TITLE
Add Library Panel Resource

### PR DESCRIPTION
**Pull request by justinTM**
This will be my first substantial Go code, and it also relies on some yet-nonexistent grafana go API additions for library panels

### CODE REVIEW

**Comment by justinTM** [comment]
@julienduchesne should i remove the `replace ...` now or after the review?

**Comment by julienduchesne** [comment]
> @julienduchesne should i remove the `replace ...` now or after the review?

Didn't realize that this was ready to test due to the failing tests (not your fault, library panels don't exist in v7). I can take a look tomorrow. I'll probably push something to your branch to ignore this resource on the v7 tests

To answer your question, after is fine. When I approve this PR, I'll merge the client and then we can update this PR and merge it too

File: grafana/resource_library_panel.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+			"updated": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Unique ID (UID) of the folder containing the library panel.",
+			},
+			// "connected_dashboards": {
+			// 	Type:     	 	schema.TypeList,
+			// 	Computed:    	true,
+			// 	Description: 	"Unique ID (UID) of the folder containing the library panel.",
+			// },
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go]: I'd remove this for now

File: grafana/resource_library_panel.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+			},
+			"folder_uid": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Unique ID (UID) of the folder containing the library panel.",
+			},
+			"created": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Unique ID (UID) of the folder containing the library panel.",
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go]: Copy pasted description

File: grafana/resource_library_panel.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+			},
+			"updated": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Unique ID (UID) of the folder containing the library panel.",
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go]: Copy pasted description

File: grafana/resource_library_panel.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		UpdateContext: UpdateLibraryPanel,
+		DeleteContext: DeleteLibraryPanel,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"uid": {
+				Type:     schema.TypeString,
+				Computed: true,
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go:36]: I believe you need to set `Optional: true` as well if you want the UID to be settable

File: grafana/resource_library_panel.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	d.SetId(resp.UID)
+	d.Set("uid", resp.UID)
+	return ReadLibraryPanel(ctx, d, meta)
+}
+
+func DeleteLibraryPanel(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	uid := d.Id()
+	_, err := client.DeleteLibraryPanel(uid)
+	var diags diag.Diagnostics
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go]: This is not needed. You can return `nil` at the end of the function instead

File: grafana/resource_library_panel.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+// validateLibraryPanelModelJSON is the ValidateFunc for `model_json`. It
+// ensures its value is valid JSON.
+func validateLibraryPanelModelJSON(model interface{}, k string) ([]string, []error) {
+	modelJSON := model.(string)
+	modelMap := map[string]interface{}{}
+	err := json.Unmarshal([]byte(modelJSON), &modelMap)
+	if err != nil {
+		return nil, []error{err}
+	}
+	return nil, nil
```
Comment (julienduchesne) [inline grafana/resource_library_panel.go]: You can reuse `unmarshalLibraryPanelModelJSON` here
```suggestion
	if _, err := unmarshalLibraryPanelModelJSON(modelJSON); err != nil {
		return nil, []error{err}
	}
	return nil, nil
```

File: grafana/resource_library_panel_test.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	"testing"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccLibraryPanel_basic(t *testing.T) {
+	CheckOSSTestsEnabled(t)
```
Comment (julienduchesne) [inline grafana/resource_library_panel_test.go:14]: You can merge in master and add the following line to ignore v7 tests
```suggestion
	CheckOSSTestsEnabled(t)
	CheckOSSTestsSemver(t, ">=8.0.0")
```

File: go.mod [hunk 8]
```diff
@@ -12,3 +12,5 @@ require (
 	github.com/hashicorp/terraform-plugin-docs v0.5.1
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.10.1
 )
+
+replace github.com/grafana/grafana-api-golang-client v0.2.4 => github.com/justinTM/grafana-api-golang-client v0.2.2-0.20220127052552-a87f7a60db01
```
Comment (julienduchesne) [inline go.mod]: I've merged your PR and released v0.2.5. You can update this

**Review by julienduchesne — COMMENTED** [review]
Looking very good! Just a few comments and we should be able to merge this!

**Comment by justinTM** [comment]
@julienduchesne good to go! Thank you for your review

**Review by julienduchesne — APPROVED** [review]
LGTM! Thank you for your contribution @justinTM!

**Pull request merged**